command/013_config_upgrade.go

```diff
@@ -71,58 +75,386 @@ func (c *ZeroThirteenUpgradeCommand) Run(args []string) int {
+	for _, path := range primary {
+		file, fileDiags := parser.LoadConfigFile(path)
+		diags = diags.Append(fileDiags)
+		if file != nil {
+			files[path] = file
 		}
 	}
-
-	// Find the provider dependencies
-	configDeps, depsDiags := earlyConfig.ProviderDependencies()
-	if depsDiags.HasErrors() {
-		c.Ui.Error(strings.TrimSpace("Could not detect provider dependencies"))
-		diags = diags.Append(depsDiags)
+	if diags.HasErrors() {
+		c.Ui.Error(strings.TrimSpace("Failed to load configuration"))
 		c.showDiagnostics(diags)
 		return 1
 	}
 
-	// Detect source for each provider
-	providerSources, detectDiags := detectProviderSources(configDeps.Providers)
-	if detectDiags.HasErrors() {
-		c.Ui.Error(strings.TrimSpace("Unable to detect sources for providers"))
-		diags = diags.Append(detectDiags)
-		c.showDiagnostics(diags)
-		return 1
+	// FIXME: It's not clear what the correct behaviour is for upgrading
+	// override files. For now, just log that we're ignoring the file.
+	for _, path := range overrides {
+		c.Ui.Warn(fmt.Sprintf("Ignoring override file %q: not implemented", path))
 	}
 
-	if len(providerSources) == 0 {
-		c.Ui.Output("No non-default providers found. Your configuration is ready to use!")
-		return 0
+	// Build up a list of required providers, uniquely by local name
+	requiredProviders := make(map[string]*configs.RequiredProvider)
+	var rewritePaths []string
+
+	// Step 1: copy all explicit provider requirements across
+	for path, file := range files {
+		for _, rps := range file.RequiredProviders {
+			rewritePaths = append(rewritePaths, path)
+			for _, rp := range rps.RequiredProviders {
+				if previous, exist := requiredProviders[rp.Name]; exist {
+					diags = diags.Append(&hcl.Diagnostic{
+						Summary:  "Duplicate required provider configuration",
+						Detail:   fmt.Sprintf("Found duplicate required provider configuration for %q.Previously configured at %s", rp.Name, previous.DeclRange),
+						Severity: hcl.DiagWarning,
+						Context:  rps.DeclRange.Ptr(),
+						Subject:  rp.DeclRange.Ptr(),
+					})
+				} else {
+					// We're copying the struct here to ensure that any
+					// mutation does not affect the original, if we rewrite
+					// this file
+					requiredProviders[rp.Name] = &configs.RequiredProvider{
+						Name:        rp.Name,
+						Source:      rp.Source,
+						Type:        rp.Type,
+						Requirement: rp.Requirement,
+						DeclRange:   rp.DeclRange,
+					}
+				}
+			}
+		}
+	}
+
+	for _, file := range files {
+		// Step 2: add missing provider requirements from provider blocks
+		for _, p := range file.ProviderConfigs {
+			// If no explicit provider configuration exists for the
+			// provider configuration's local name, add one with a legacy
+			// provider address.
+			if _, exist := requiredProviders[p.Name]; !exist {
+				requiredProviders[p.Name] = &configs.RequiredProvider{
+					Name:        p.Name,
+					Type:        addrs.NewLegacyProvider(p.Name),
+					Requirement: p.Version,
```

Comment:
I have some reservations about including the provider version from the `provider`  to `provider_requirements`. Personally I'd find it offputting to end up with duplicate version constraints in the configs. What do you think? (this does not need to block merging)

```hcl
# main.tf
provider "foo" {
    version = "1.0.0"
}
```

```hcl
# providers.tf
terraform {
   required_providers {
      foo = {
          source = "hashicorp/foo"
          version = "1.0.0"
      }
   }
}
```

Reply:
Oh, good call! I didn't spot that on my review.

I think I have a slightly more controversial take on this, though: what if we were to move the constraints into the `required_providers` block and _remove_ the now-duplicated arguments inside `provider` blocks?

(The version constraints inside `provider` blocks have been quietly deprecated since 0.12.0 anyway because it doesn't behave well in non-root modules, so we could take this opportunity to help folks move to the new intended model so that it's less disruptive to make this deprecation more noisy somewhere down the line...)

Reply:
Good catch, @mildwonkey, thanks!

For now, I've taken the quicker route of not duplicating the provider version constraint. I think it's worth considering rewriting the `provider` blocks, but it's not trivial so I don't want to bundle it with this PR. I've taken a note to revisit.